Black and White listings for the Extension Manager (#7989)
jupyterlab/jupyterlab: docs/source/user/extensions.rst

Listings
~~~~~~~~

When searching extensions, JupyterLab displays the complete search result and 

> the user if free to install any extension. This is the :ref:`default_mode`.

Reviewer: ```suggestion
the user is free to install any extension. This is the :ref:`default_mode`.
```

To bring more security, you or your administrator can enable ``blacklists`` or ``whitelists``
mode. JupyterLab will check the extensions against the defined listings.

Only one mode at a time is allowed.

The following details the behavior for the :ref:`blacklist_mode` and the :ref:`whitelist_mode`.
The details to enable configure the listings can be read :ref:`listings_conf`. 

.. _default_mode:

Default mode
^^^^^^^^^^^^

In the ``default`` mode, no listing is enabled and the search behavior is unchanged and

> is the one described previously.

Reviewer: ```suggestion
In the ``default`` mode, no listing is enabled and the search behavior is unchanged,
as described previously.
```

Author: Committed. Thx.

However, users can add malicious extensions to a blacklist. The extension manager 
will show all extensions except for those that have 
been explicitly added to the blacklist. Therfore, the extension manager 
does not allow you to install blacklisted extensions.

If you, or your administrator, has enabled the blacklist mode,
JupyterLab will use the blacklist and remove all blacklisted
extensions from your search result.

If you have installed an extension before it has been blacklisted,
the extension entry in the installed list will be highlighted
in red. It is recommended that you uninstall it. You can move
your mouse on the question mark icon to read the instructions.

.. figure:: images/listings/installed_blacklisted.png
   :align: center
   :class: jp-screenshot

   **Figure:** Blacklisted installed extension which should be removed

.. _whitelist_mode:

Whitelist mode
^^^^^^^^^^^^^^

A whitelist maintains a set of approved extensions that users can freely 
search and install. Extensions need to go through some sort of vetting process 
before they are added to the whitelist. When using a whitelist, the extension manager 

> will highlight extensions that have been explicitly added to the whitelist.

Reviewer: ```suggestion
will only show extensions that have been explicitly added to the whitelist.
```

If you, or your administrator, has enabled the whitelist mode
JupyterLab will use the whitelist and only show extensions present

> in the withelist. The other extensions will not be show in the search result.

Reviewer: ```suggestion
in the whitelist. The other extensions will not be show in the search result.
```

Author: Committed. Thx.

.. _listings_conf:

Listing Configuration
^^^^^^^^^^^^^^^^^^^^^

You or your administrator can use the following traits to define the listings loading.

- ``blacklist_uris``: A list of comma-separated URIs to get the blacklist
- ``whitelist_uris``: A list of comma-separated URIs to get the whitelist
- ``listings_refresh_ms``: The interval delay in milliseconds to refresh the lists
- ``listings_request_options``: The optional kwargs to use for the listings HTTP requests

For example, to enable blacklist, launch the server with ``--LabServerApp.blacklist_uris``.

The details for the listings_request_options are listed
on the `this page <https://2.python-requests.org/en/v2.7.0/api/#requests.request>`__  
(for example, you could pass ``{'timeout': 10}`` to change the HTTP request timeout value).

The listings are json files hosted on the URIs you have given.

This is an example of a blacklist file.

.. code:: json

   {
   "blacklist": [
      {
         "name": "@jupyterlab-examples/launcher",
         "type": "jupyterlab",
         "reason": "@jupyterlab-examples/launcher is blacklisted for test purpose - Do NOT take this for granted!!!",

>          "creation_date": "2020-03-11T03:28:56.782Z",

Reviewer: Thoughts on removing all keys besides `name` or making them optional?

Just wondering what they add to the UI at the moment and wanting to ensure it's as easy as possible to maintain these files.

Author: Made the name mandatory and the others optional. Also documented that.
